registry/storage/driver/speedy/speedy.go

```diff
@@ -0,0 +1,569 @@
+func (d *driver) ReadStream(ctx context.Context, path string, offset int64) (io.ReadCloser, error) {
+	//get all metainfos
+	url, err := d.getURL()
+	if err != nil {
+		return nil, err
+	}
+
+	infoArr, err := d.client.GetFileInfo(url, path)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(infoArr) == 0 {
+		return nil, fmt.Errorf("There is not file info, path: %v", path)
+	}
+
+	return d.initReadStream(path, uint64(offset), infoArr)
+}
+
+func (d *driver) initReadStream(path string, offset uint64, infoArr []*MetaInfoValue) (io.ReadCloser, error) {
+	var size uint64
+	for index := range infoArr {
+		size += infoArr[index].End - infoArr[index].Start
+	}
+
+	if uint64(offset) > size {
+		return nil, storagedriver.InvalidOffsetError{Path: path, Offset: int64(offset)}
+	}
+
+	readCloser := &readStreamReader{
+		driver:       d,
+		path:         path,
+		index:        0,
+		infoArr:      infoArr,
+		remain:       nil,
+		remainOffset: 0,
+		size:         size,
+		offset:       0,
+	}
+
+	var index int
+	var readOffset uint64
+	for index < len(infoArr) {
+		nextReadOffset := readOffset + (infoArr[index].End - infoArr[index].Start)
+		if nextReadOffset >= uint64(offset) {
+			break
+		}
+		readOffset = nextReadOffset
+		index++
+		continue
+	}
+
+	url, err := d.getURL()
+	if err != nil {
+		return nil, err
+	}
+
+	readCloser.remain, err = d.client.DownloadFile(url, path, infoArr[index])
+	if err != nil {
+		return nil, err
+	}
+	readCloser.index = index + 1
+	readCloser.remainOffset = offset - readOffset
+	readCloser.offset = offset
+	return readCloser, nil
+}
+
+// WriteStream stores the contents of the provided io.ReadCloser at a
+// location designated by the given path.
+// May be used to resume writing a stream by providing a nonzero offset.
+// The offset must be no larger than the CurrentSize for this path.
+func (d *driver) WriteStream(ctx context.Context, path string, offset int64, reader io.Reader) (int64, error) {
+	url, err := d.getURL()
+	if err != nil {
+		return 0, err
+	}
+
+	infoArr, err := d.client.GetFileInfo(url, path)
+	if err != nil {
+		return 0, err
+	}
+
+	var totalRead int64
+	var index int
+	var currentSize uint64
+
+	//Align to chunk size, skip already exist chunk
+	if len(infoArr) != 0 {
+		for _, info := range infoArr {
+			currentSize += info.End - info.Start
+			index++
+		}
+
+		if currentSize < uint64(offset) {
+			return 0, fmt.Errorf("speedy driver currentSize: %d < offset: %d", currentSize, offset)
+		}
+
+		//read 1MB every time
+		var tempSize uint64 = 1 << 20
+		tempBuf := make([]byte, tempSize)
+		var distance uint64
+		if currentSize > uint64(offset) {
+			distance = currentSize - uint64(offset)
+		}
+
+		for distance > 0 {
+			var readSize uint64
+			if distance > tempSize {
+				readSize = tempSize
+			} else {
+				readSize = distance
+			}
+
+			n, err := reader.Read(tempBuf[0:readSize])
+			if err != nil {
+				return totalRead, err
```

Comment:
in registry/storage/filewriter.go
// Write writes the buffer p at the current write offset.
func (fw *fileWriter) Write(p []byte) (n int, err error) {
    nn, err := fw.ReadFrom(bytes.NewReader(p))
    return int(nn), err  
} 


// ReadFrom reads reader r until io.EOF writing the contents at the current
// offset.
func (fw *fileWriter) ReadFrom(r io.Reader) (n int64, err error) {
    if fw.err != nil {   
        return 0, fw.err 
    }   
    
    nn, err := fw.driver.WriteStream(fw.ctx, fw.path, fw.offset, r) //the problem about totoalRead+n is in Speedy.WriteStream
    
    // We should forward the offset, whether or not there was an error.
    // Basically, we keep the filewriter in sync with the reader's head. If an
    // error is encountered, the whole thing should be retried but we proceed
    // from an expected offset, even if the data didn't make it to the 
    // backend.
    fw.offset += nn
    
    if fw.offset > fw.size {   
        fw.size = fw.offset    
    }   


    return nn, err 
} 


go/src/pkg/bytes/reader.go     
func (r *Reader) Read(b []byte) (n int, err error) {
    if len(b) == 0 { 
        return 0, nil 
    }
    if r.i >= int64(len(r.s)) {
        return 0, io.EOF  //we can find that when err is not nil,  n == 0
    }
    r.prevRune = -1
    n = copy(b, r.s[r.i:])
    r.i += int64(n)
    return
}


if err != nil, n == 0, so returning totalRead do not need add n.


--


At 2015-08-12 06:28:25, "Anton Tyurin" <notifications@github.com> wrote:


In registry/storage/driver/speedy/speedy.go:

> +		var distance uint64
> +		if currentSize > uint64(offset) {
> +			distance = currentSize - uint64(offset)
> +		}
> +
> +		for distance > 0 {
> +			var readSize uint64
> +			if distance > tempSize {
> +				readSize = tempSize
> +			} else {
> +				readSize = distance
> +			}
> +
> +			n, err := reader.Read(tempBuf[0:readSize])
> +			if err != nil {
> +				return totalRead, err


n can be > 0, so returning totalRead withoud adding n is not correct.

—
Reply to this email directly or view it on GitHub.